File: Code/Avernakis Nodejs/Avernakis-Nodejs/Private/UiRichListBox.cpp
#include "StdAfx.h"
#include "UiRichListBox.h"
#include "UiHeader.h"

#define ThisMethod($x) &UiRichListBox::$x
#define AutoAddMethod($x, ...) AddMethod<__VA_ARGS__>( #$x, ThisMethod( $x ) )
#define MakeThisFunc($x) MakeFunc( this, &UiRichListBox::$x )

namespace Nav
{
	
	namespace
	{
		ObjectRegister<UiRichListBox> c_obj;
	}

	void UiRichListBox::DefineControl()
	{
		AutoAddMethod( ItemInsert );
		AutoAddMethod( ItemRemove );
		AutoAddMethod( ItemClear );
		AutoAddMethod( ItemSelect );
		AutoAddMethod( ItemSetSelect );
		AutoAddMethod( ItemSetSelectRange );
		AutoAddMethod( ItemSort );
		AutoAddMethod( ItemFind, WrapObjectGeneric );
		AutoAddMethod( ItemSet );
		AutoAddMethod( ItemGet, WrapObjectGeneric );
		AutoAddMethod( ItemGetCount, WrapObjectGeneric );
		AutoAddMethod( ItemGetCountPerPage, WrapObjectGeneric );
		AutoAddMethod( ItemGetSelectionCount, WrapObjectGeneric );
		AutoAddMethod( ItemGetSelection, WrapObjectGeneric );
		AutoAddMethod( ItemIsSelected, WrapObjectGeneric );
		AutoAddMethod( ItemGetNextSelection, WrapObjectGeneric );
		AutoAddMethod( ItemEnsure );
		AutoAddMethod( ItemGetTopIndex, WrapObjectGeneric );
		AutoAddMethod( ItemIsVisible, WrapObjectGeneric );
		AutoAddMethod( ItemGetRect, WrapObjectGeneric );
		AutoAddMethod( ItemHitTest, WrapObjectGeneric );
		AutoAddMethod( ItemSetCount );
		AutoAddMethod( ItemSetHighlight );
		AutoAddMethod( ItemSetMinimumHeight );
		AutoAddMethod( ItemGetMinimumHeight );

		AutoAddMethod( SetVirtual );

		AutoAddMethod( GetHeader, WrapObjectGeneric );
		AutoAddMethod( SetHeaderAutoWidth );
		AutoAddMethod( GetHeaderAutoWidth, WrapObjectGeneric );

		AutoAddMethod( SetScrollPosition );
		AutoAddMethod( GetScrollPosition, WrapObjectGeneric );
		AutoAddMethod( GetScrollSize, WrapObjectGeneric );
		AutoAddMethod( GetScrollMax, WrapObjectGeneric );

		AutoAddMethod( SetSelectionMode );
		AutoAddMethod( GetSelectionMode, WrapObjectGeneric );

		AutoAddMethod( SetViewMode );
		AutoAddMethod( GetViewMode, WrapObjectGeneric );

		AutoAddMethod( SetTileSize );
		AutoAddMethod( GetTileSize, WrapObjectGeneric );
		AutoAddMethod( SetTileColumnPriority );
		AutoAddMethod( GetTileColumnPriority, WrapObjectGeneric );

		AutoAddMethod( SetItemSpace );
		AutoAddMethod( GetItemSpace, WrapObjectGeneric );

		AutoAddMethod( SetTreeStyle );
		AutoAddMethod( GetTreeStyle, WrapObjectGeneric );

		AutoAddMethod( SetItemCheck );
		AutoAddMethod( GetItemCheck, WrapObjectGeneric );

		AutoAddMethod( SetIconSize );
		AutoAddMethod( GetIconSize, WrapObjectGeneric );

		AutoAddMethod( SetIconNameColumn );
		AutoAddMethod( GetIconNameColumn, WrapObjectGeneric );

		AutoAddMethod( SetListItemWidth );
		AutoAddMethod( GetListItemWidth, WrapObjectGeneric );

		AutoAddMethod( GetListRect, WrapObjectGeneric );

		AutoAddMethod( OnSelectionChange /**/, WrapObjectGeneric );
		AutoAddMethod( OnSelectionEnd    /**/, WrapObjectGeneric );
		AutoAddMethod( OnDoubleClick     /**/, WrapObjectGeneric );
		AutoAddMethod( OnRightClick      /**/, WrapObjectGeneric );
		AutoAddMethod( OnDragBegin       /**/, WrapObjectGeneric );
		AutoAddMethod( OnGlyphClick      /**/, WrapObjectGeneric );
		AutoAddMethod( OnItemCheck       /**/, WrapObjectGeneric );
	}

	U1 UiRichListBox::Ctor( const CallbackInfo& ci, UiWindow * p, Napi::Value v )
	{
		if ( !__CreateControl( p, v ) )
			return false;

		m_Header = ci.NewJsObjectWithOwnership<UiHeader>();
		if ( !m_Header )
			return false;
		m_Header->SetSharedControl( &GetControlTyped().GetHeader() );

		GetControlTyped().SetIconManager( Ui::RichListBoxViewMode::Detail, p->PublicCloneIconManager() );
		GetControlTyped().SetIconManager( Ui::RichListBoxViewMode::Tile, p->PublicCloneIconManager() );
		GetControlTyped().SetIconManager( Ui::RichListBoxViewMode::Icon, p->PublicCloneIconManager() );
		GetControlTyped().SetIconManager( Ui::RichListBoxViewMode::List, p->PublicCloneIconManager() );

		return true;
	}

	void UiRichListBox::__OnSelectionChange( Ui::IRichListBox & sender )
	{
		m_OnSelectionChange( this );
	}

	void UiRichListBox::__OnSelectionEnd( Ui::IRichListBox & sender )
	{
		m_OnSelectionEnd( this );
	}

	void UiRichListBox::__OnDoubleClick( Ui::IRichListBox & sender, S32 nIndex )
	{
		m_OnDoubleClick( this, nIndex );
	}

	void UiRichListBox::__OnRightClick( Ui::IRichListBox & sender, S32 nIndex )
	{
		m_OnRightClick( this, nIndex );
	}

	void UiRichListBox::__OnDragBegin( Ui::IRichListBox & sender )
	{
		m_OnDragBegin( this );
	}

	void UiRichListBox::__OnGlyphClick( Ui::IRichListBox & sender, S32 nIndex )
	{
		m_OnGlyphClick( this, nIndex );
	}

	void UiRichListBox::__OnItemCheck( Ui::IRichListBox & sender, S32 nIndex )
	{
		m_OnItemCheck( this, nIndex );
	}

	void UiRichListBox::__OnVirtual( Ui::IRichListBox & sender, Ui::RichListBoxItemVirtual & iv )
	{
		//if ( !App::GetSingleton().BlockCallTest() )
		//{
		//	auto r = m_Virtual.DirectCall( this, iv.m_Item, iv.m_SubItem );
		//	iv.m_Icon = r.m_Icon;
		//	iv.m_Indent = r.m_Indent;
		//	iv.m_Misc = r.m_Misc;
		//	iv.m_StringData = r.m_String;
		//	return;
		//}
		UiRichListBoxItemVirtual_t r;
		m_Virtual.BlockCall( this, iv.m_Item, iv.m_SubItem, r );
		iv.m_Icon = r.m_Icon;
		iv.m_Indent = r.m_Indent;
		iv.m_Misc = r.m_Misc;
		iv.m_StringData = r.m_String;
	}

	S32 UiRichListBox::ItemInsert( const UiRichListBoxItem_t & pItem, U1 bReserveSelection )
	{
		Ui::RichListBoxItem item{};
		item.m_Flag = pItem.m_Flag;
		item.m_Index = pItem.m_Index;
		item.m_IndexSub = pItem.m_IndexSub;
		item.m_Icon = pItem.m_Icon;
		item.m_Text = (PWChar) pItem.m_Text.c_str();
		item.m_Indent = pItem.m_Indent;
		item.m_GroupId = pItem.m_GroupId;
		item.m_Misc = pItem.m_Misc;
		item.m_Context = (void*) (USize) pItem.m_Context;
		return GetControlTyped().ItemInsert( item, bReserveSelection );
	}

	U1 UiRichListBox::ItemSet( const UiRichListBoxItem_t & pItem )
	{
		Ui::RichListBoxItem item{};
		item.m_Flag = pItem.m_Flag;
		item.m_Index = pItem.m_Index;
		item.m_IndexSub = pItem.m_IndexSub;
		item.m_Icon = pItem.m_Icon;
		item.m_Text = (PWChar) pItem.m_Text.c_str();
		item.m_Indent = pItem.m_Indent;
		item.m_GroupId = pItem.m_GroupId;
		item.m_Misc = pItem.m_Misc;
		item.m_Context = (void*) (USize) pItem.m_Context;
		return GetControlTyped().ItemSet( item );
	}

	UiRichListBoxItem_t UiRichListBox::ItemGet( U32 nFlag, S32 nIndex, S32 nSubIndex )
	{
		Ui::RichListBoxItem item{ nFlag, nIndex, nSubIndex };
		WChar szText[1024]{};
		item.m_Text = szText;
		item.m_TextCharLength = CountOf( szText );

		UiRichListBoxItem_t r{};
		r.m_Flag = nFlag;
		r.m_Index = nIndex;
		r.m_IndexSub = nSubIndex;
		if ( GetControlTyped().ItemGet( item ) )
		{
			r.m_Icon = item.m_Icon;
			r.m_Text = szText;
			r.m_Indent = item.m_Indent;
			r.m_GroupId = item.m_GroupId;
			r.m_Misc = item.m_Misc;
			r.m_Context = (U32) (USize) item.m_Context;
		}
		return r;
	}

	UiRichListBoxHitTestResult_t UiRichListBox::ItemHitTest( const WrapData<S32_2>& pt, U1 bMustOnContent, U1 bIncludePlaceholder ) const
	{
		UiRichListBoxHitTestResult_t r{};
		if ( bMustOnContent )
			r.m_Index = GetControlTyped().ItemHitTest( pt, r.m_SubIndex, bIncludePlaceholder, &r.m_Place, &r.m_PlaceAb );
		else
			r.m_Index = GetControlTyped().ItemHitTest( pt, false, bIncludePlaceholder, &r.m_Place, &r.m_PlaceAb );
		return r;
	}

	UiRichListBox * UiRichListBox::SetVirtual( Virtual_t && fn )
	{
		m_Virtual = std::move( fn );
		if ( m_Virtual )
			GetControlTyped().SetVirtual( MakeThisFunc( __OnVirtual ) );
		else
			GetControlTyped().SetVirtual( {} );
		return this;
	}

	WrapData<S32_2> UiRichListBox::GetScrollMax() const
	{
		S32_2 v{};
		if ( GetControlTyped().GetScrollH().GetVisible() )
			v.x = GetControlTyped().GetScrollH().GetMaximum();
		if ( GetControlTyped().GetScrollV().GetVisible() )
			v.y = GetControlTyped().GetScrollV().GetMaximum();
		return v;
	}

	UiRichListBox* UiRichListBox::OnSelectionChange /**/( OnCallback0_t&& fn ) { m_OnSelectionChange /**/ = SetEventCallback<Ui::IRichListBox::OnSelectionChange /**/>( std::move( fn ), MakeThisFunc( __OnSelectionChange /**/ ) ); return this; }
	UiRichListBox* UiRichListBox::OnSelectionEnd    /**/( OnCallback0_t&& fn ) { m_OnSelectionEnd    /**/ = SetEventCallback<Ui::IRichListBox::OnSelectionEnd    /**/>( std::move( fn ), MakeThisFunc( __OnSelectionEnd    /**/ ) ); return this; }
	UiRichListBox* UiRichListBox::OnDoubleClick     /**/( OnCallback1_t&& fn ) { m_OnDoubleClick     /**/ = SetEventCallback<Ui::IRichListBox::OnDoubleClick     /**/>( std::move( fn ), MakeThisFunc( __OnDoubleClick     /**/ ) ); return this; }
	UiRichListBox* UiRichListBox::OnRightClick      /**/( OnCallback1_t&& fn ) { m_OnRightClick      /**/ = SetEventCallback<Ui::IRichListBox::OnRightClick      /**/>( std::move( fn ), MakeThisFunc( __OnRightClick      /**/ ) ); return this; }
	UiRichListBox* UiRichListBox::OnDragBegin       /**/( OnCallback0_t&& fn ) { m_OnDragBegin       /**/ = SetEventCallback<Ui::IRichListBox::OnDragBegin       /**/>( std::move( fn ), MakeThisFunc( __OnDragBegin       /**/ ) ); return this; }
	UiRichListBox* UiRichListBox::OnGlyphClick      /**/( OnCallback1_t&& fn ) { m_OnGlyphClick      /**/ = SetEventCallback<Ui::IRichListBox::OnGlyphClick      /**/>( std::move( fn ), MakeThisFunc( __OnGlyphClick      /**/ ) ); return this; }
	UiRichListBox* UiRichListBox::OnItemCheck       /**/( OnCallback1_t&& fn ) { m_OnItemCheck       /**/ = SetEventCallback<Ui::IRichListBox::OnItemCheck       /**/>( std::move( fn ), MakeThisFunc( __OnItemCheck       /**/ ) ); return this; }
}

File: Code/Avernakis Nodejs/Test-Nodejs/src/Ave/Ui/Control/UiRichListBox.ts
import { IControl } from "../UiControl";
import { AveLib, ExtendControlInstance } from "../../AveLib";
import { Window } from "./UiWindow";
import { DpiSize, DpiSize_2, IconCache, Rect } from "../UiCommon";
import { Vec2 } from "../../Math/Vector";
import { Header } from "./UiHeader";

export enum RichListBoxSelectionMode {
    Single,
    Multiple,
}

export enum RichListBoxViewMode {
    Detail,
    Tile,
    Icon,
    List,
}

export enum RichListBoxItemPlace {
    Item,
    Above,
    Below,
}

export enum RichListBoxItemCheck {
    None,
    Selection,
    Standalone,
}

export enum RichListBoxItemFlag {
    None /**/ = 0x0000,
    Text /**/ = 0x0001,
    Indent /**/ = 0x0002,
    GroupId /**/ = 0x0004,
    Context /**/ = 0x0008,
    Icon /**/ = 0x0010,
    Misc /**/ = 0x0020,
}

export enum RichListBoxItemMisc {
    None /**/ = 0x0000, // Draw a tree glyph
    HasChild /**/ = 0x0001, // Draw a tree glyph
    Expanded /**/ = 0x0002, // Draw an expanded tree glyph
    NoChild /**/ = 0x0004, // The item has no glyph (no glyph indent)
    PlaceHolder /**/ = 0x0008,
    Checked /**/ = 0x0010, // Only valid in RichListBoxItemCheck::Standalone mode
    CheckHidden /**/ = 0x0020, // Only valid in RichListBoxItemCheck::Standalone mode
    SingleLine /**/ = 0x0100, // Single line text, only for tile mode
}

export class RichListBoxItem {
    Flag: RichListBoxItemFlag = RichListBoxItemFlag.None;
    Index: number = 0;
    IndexSub: number = 0;
    Icon: IconCache = 0;
    Text: string = "";
    Indent: number = 0;
    GroupId: number = 0;
    Misc: RichListBoxItemMisc = RichListBoxItemMisc.None;
    Context: number = 0;
}

export class RichListBoxHitTestResult {
    Index: number;
    SubIndex: number;
    Place: RichListBoxItemPlace;
    PlaceAb: RichListBoxItemPlace;
}

export class RichListBoxItemVirtual {
    Icon: IconCache = 0;
    Indent: number = 0;
    Misc: RichListBoxItemMisc = RichListBoxItemMisc.None;
    String: string = "";
}

export interface IRichListBox extends IControl {
    new (window: Window): IRichListBox;

    ItemInsert(item: RichListBoxItem, bReserveSelection: boolean): number;
    ItemRemove(nIndex: number, bReserveSelection: boolean): boolean;
    ItemClear(): RichListBox;
    ItemSelect(nIndex: number): RichListBox;
    ItemSetSelect(nIndex: number, bSelect: boolean): RichListBox;
    ItemSetSelectRange(
        nStart: number,
        nEnd: number,
        bSelect: boolean
    ): RichListBox; // in the range [nStart, nEnd], not [nStart, nEnd)
    ItemSort(
        nSubItem: number,
        bAsc: boolean,
        bCaseInsensitive: boolean
    ): RichListBox;
    ItemFind(
        szString: string,
        nIndex: number,
        bExact: boolean,
        bCaseInsensitive: boolean
    ): number;
    ItemSet(item: RichListBoxItem): boolean;
    ItemGet(
        nFlag: RichListBoxItemFlag,
        nIndex: number,
        nSubIndex: number
    ): RichListBoxItem;
    ItemGetCount(): number;
    ItemGetCountPerPage(bIncludePartial: boolean): number;
    ItemGetSelectionCount(): number;
    ItemGetSelection(): number;
    ItemIsSelected(nIndex: number): boolean;
    ItemGetNextSelection(nIndex: number): number;
    ItemEnsure(nIndex: number): RichListBox;
    ItemGetTopIndex(): number;
    ItemIsVisible(nIndex: number): boolean;
    ItemGetRect(nIndex: number): Rect;
    ItemHitTest(
        pt: Vec2,
        bMustOnContent: boolean,
        bIncludePlaceholder: boolean
    ): RichListBoxHitTestResult;
    ItemSetCount(nCount: number): RichListBox;
    ItemSetHighlight(nIndex: number, nPlace: RichListBoxItemPlace): RichListBox;
    ItemSetMinimumHeight(nHeight: DpiSize): RichListBox;
    ItemGetMinimumHeight(): DpiSize;

    SetVirtual(
        fn: (
            sender: RichListBox,
            nIndex: number,
            nSubIndex: number
        ) => RichListBoxItemVirtual
    ): RichListBox;

    GetHeader(): Header;
    SetHeaderAutoWidth(b: boolean): RichListBox;
    GetHeaderAutoWidth(): boolean;

    SetScrollPosition(vPos: Vec2): RichListBox;
    GetScrollPosition(): Vec2;

    SetSelectionMode(nSelMode: RichListBoxSelectionMode): RichListBox;
    GetSelectionMode(): RichListBoxSelectionMode;

    SetViewMode(nViewMode: RichListBoxViewMode): RichListBox;
    GetViewMode(): RichListBoxViewMode;

    SetTileSize(vSize: DpiSize_2): RichListBox;
    GetTileSize(): DpiSize_2;
    SetTileColumnPriority(order: number[]): RichListBox;
    GetTileColumnPriority(): Array<number>;

    SetItemSpace(vSize: DpiSize_2): RichListBox;
    GetItemSpace(): DpiSize_2;

    SetTreeStyle(b: boolean): RichListBox;
    GetTreeStyle(): boolean;

    SetItemCheck(nSelMode: RichListBoxItemCheck): RichListBox;
    GetItemCheck(): RichListBoxItemCheck;

    SetIconSize(nMode: RichListBoxViewMode, vSize: DpiSize): RichListBox;
    GetIconSize(nMode: RichListBoxViewMode): DpiSize;

    SetIconNameColumn(nColumn: number): RichListBox;
    GetIconNameColumn(): number;

    SetListItemWidth(nColumn: DpiSize): RichListBox;
    GetListItemWidth(): DpiSize;

    GetListRect(): Rect;

    OnSelectionChange(/**/ fn: (sender: RichListBox) => void): RichListBox;
    OnSelectionEnd(/**/ fn: (sender: RichListBox) => void): RichListBox;
    OnDoubleClick(
        /**/ fn: (sender: RichListBox, nIndex: number) => void
    ): RichListBox;
    OnRightClick(
        /**/ fn: (sender: RichListBox, nIndex: number) => void
    ): RichListBox;
    OnDragBegin(/**/ fn: (sender: RichListBox) => void): RichListBox;
    OnGlyphClick(
        /**/ fn: (sender: RichListBox, nIndex: number) => void
    ): RichListBox;
    OnItemCheck(
        /**/ fn: (sender: RichListBox, nIndex: number) => void
    ): RichListBox;
}

export class RichListBox extends (AveLib.UiRichListBox as IRichListBox) {
    ItemGetRect(nIndex: number): Rect {
        return Rect.FromNative(super.ItemGetRect(nIndex));
    }

    GetListRect(): Rect {
        return Rect.FromNative(super.GetListRect());
    }

    GetScrollPosition(): Vec2 {
        return Vec2.FromNative(super.GetScrollPosition());
    }

    GetHeader(): Header {
        const header = super.GetHeader();
        ExtendControlInstance(header as any as IControl);
        return header;
    }
}
